k256/src/succinct.rs

```diff
@@ -0,0 +1,682 @@
+
+        fn y_is_odd(&self) -> Choice {
+            let (_, y) = self.field_elements();
+
+            Choice::from(y.normalize().is_odd())
+        }
+    }
+
+    impl ConditionallySelectable for Sp1AffinePoint {
+        fn conditional_select(a: &Self, b: &Self, choice: Choice) -> Self {
+            // In the vm, we dont care about constant time selection.
+            if choice.into() {
+                *b
+            } else {
+                *a
+            }
+        }
+    }
+
+    impl ConstantTimeEq for Sp1AffinePoint {
+        fn ct_eq(&self, other: &Self) -> Choice {
+            let (x1, y1) = self.field_elements();
+            let (x1, y1) = (x1.normalize(), y1.normalize());
+
+            let (x2, y2) = other.field_elements();
+            let (x2, y2) = (x2.normalize(), y2.normalize());
+
+            x1.ct_eq(&x2) & y1.ct_eq(&y2)   
+        }
+    }
+
+    impl PartialEq for Sp1AffinePoint {
+        fn eq(&self, other: &Self) -> bool {
+            self.ct_eq(other).into()
+        }
+    }
+
+    impl Eq for Sp1AffinePoint {}
+
+    impl Default for Sp1AffinePoint {
+        fn default() -> Self {
+            Sp1AffinePoint::identity()
+        }
+    }
+
+    impl DefaultIsZeroes for Sp1AffinePoint {}
+
+    impl GroupEncoding for AffinePoint {
+        type Repr = CompressedPoint;
+    
+        fn from_bytes(bytes: &Self::Repr) -> CtOption<Self> {
+            EncodedPoint::from_bytes(bytes)
+                .map(|point| CtOption::new(point, Choice::from(1)))
+                .unwrap_or_else(|_| {
+                    // SEC1 identity encoding is technically 1-byte 0x00, but the
+                    // `GroupEncoding` API requires a fixed-width `Repr`
+                    let is_identity = bytes.ct_eq(&Self::Repr::default());
+                    CtOption::new(EncodedPoint::identity(), is_identity)
+                })
+                .and_then(|point| Self::from_encoded_point(&point))
+        }
+    
+        fn from_bytes_unchecked(bytes: &Self::Repr) -> CtOption<Self> {
+            // No unchecked conversion possible for compressed points
+            Self::from_bytes(bytes)
+        }
+    
+        fn to_bytes(&self) -> Self::Repr {
+            let encoded = self.to_encoded_point(true);
+            let mut result = CompressedPoint::default();
+            result[..encoded.len()].copy_from_slice(encoded.as_bytes());
+            result
+        }
+    }
+}
+
+/// In our case, we actually only care about affine points.
+///
+/// So this type is purely to satisfy trait bounds.
+mod projective {
+    use elliptic_curve::{group::prime::PrimeGroup, ops::MulByGenerator};
+
+    use super::*;
+
+    /// While the underlying `RustCrypto` algorithm uses projective points, we need
+    /// affine points for our syscalls.
+    ///
+    /// So this type is merely a wrapper around the `Secp256k1Point` type.
+    #[derive(Clone, Copy, Debug)]
+    pub struct Sp1ProjectivePoint {
+        inner: Sp1AffinePoint,
+    }
```

Comment:
Is there a reason why the SP1ProjectivePoint and SP1AffinePoint types can't both be in a library that the p256 and k256 patches use?